src/compiler/checker.ts

```diff
@@ -20272,18 +20272,20 @@ namespace ts {
          * file.
          */
         function isJSConstructor(node: Declaration | undefined): boolean {
-            if (node && isInJSFile(node)) {
+            if (!node || !isInJSFile(node)) {
+                return false;
+            }
+            const func = isFunctionDeclaration(node) || isFunctionExpression(node) ? node :
+                isVariableDeclaration(node) && node.initializer && isFunctionExpression(node.initializer) ? node.initializer :

```

Comment:
What about node.initializer = some Conditional expression that uses functionExpression in each branch?

Reply:
Good point, but that doesn't actually work today either, and you can put the annotation on the individual function expressions like so:

```js
const Weird = b ?
    /** @constructor */
    function () { this. x = 1 } :
    /** @constructor */
    function () { this.y = 'yes' };
```

I'd rather address that as a new feature instead of as a 3.1 bugfix.